### PR TITLE
Make getHookModuleExecList customizable from modules

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           |8.2.x
| Description?      | This PR allows modules to filter the list of modules to be invoked for a hook in the `Hook::getHookModuleExecList` method.
| Type?             | new feature
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | see below
| UI Tests          | 
| Fixed issue or discussion?     |  https://github.com/PrestaShop/PrestaShop/discussions/37060
| Related PRs       | 
| Sponsor company   | PrestaShop SA

# Explanations

With this PR, to be able to filter the list of modules returned by `Hook::getHookModuleExecList`, a module must implement a service with the tag `core.hook_module_exec_filter`, and implement the `PrestaShop\PrestaShop\Core\Hook\HookModuleFilterInterface` interface.

Inside the module, the service definition would look like this into `config/front/services.yml` :

```yml
  PrestaShop\Module\MyAwesomeModule\Service\CustomModuleList:
    tags: [ 'core.hook_module_exec_filter' ]
````

This service should be defined in `config/front/services.yml` if it's for front office purpose.

Then the service will be called automatically from `PrestaShop\PrestaShop\Core\Hook\HookModuleFilter`, which receives all services tagged as `core.hook_module_exec_filter` in its constructor.

## Test module:

Unzip in the `modules` folder, install via the module manager page in back office

In FO, check that the module `ps_linklist` doesn't display its links in the footer. 

(don't mind the name of the module, I took an existing example module as a basis)

[demodoctrine.zip](https://github.com/user-attachments/files/17958166/demodoctrine.zip)

### CODE REVIEW

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `How to test` shouldn&#039;t be empty. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#how-to-test))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!
### About linked issues
Please consider opening an issue before submitting a Pull Request:
* If it's a bug fix, it helps maintainers verify that the bug is effectively due to a defect in the code, and that it hasn't been fixed already.
* It can help trigger a discussion about the best implementation path before a single line of code is written.
* It may lead the Core Product team to mark that issue as a priority, further attracting the maintainers' attention.

(Note: this is an automated message, but answering it will reach a real human)

**Comment by kpodemski** [comment]
Interesting approach, that was actually what I suggested somewhere, to have it based on the Symfony service, but then I remembered about all kinds of issues with the Symfony container... 😅 

If I remember correctly, there's no need to add a new resource like this:
`{ resource: ../../../modules/*/config/front_services.yml }`

because if you do:

`tags: [ 'front', 'core.hook_module_exec_filter' ]`

then it should be automatically visible in the front office...?

**Comment by matthieu-rolland** [comment]
> Interesting approach, that was actually what I suggested somewhere, to have it based on the Symfony service, but then I remembered about all kinds of issues with the Symfony container... 😅
> 
> If I remember correctly, there's no need to add a new resource like this: `{ resource: ../../../modules/*/config/front_services.yml }`
> 
> because if you do:
> 
> `tags: [ 'front', 'core.hook_module_exec_filter' ]`
> 
> then it should be automatically visible in the front office...?

oh I didn't know about `tags: [ 'front', 'core.hook_module_exec_filter' ]`, I'm trying it right now !

yes, we couldn't access a symfony container so early, I had to make it fetchable from the `ContainerFinder` and get it from the Hook class...

**Comment by matthieu-rolland** [comment]
@kpodemski it doesn't seem to work, but it makes sense I guess, because the front container is a custom container that is completely distinct... or am I missing something ?

**Comment by kpodemski** [comment]
@matthieu-rolland it's been years since the last time I used this, so I have to check myself 😅

**Comment by kpodemski** [comment]
@matthieu-rolland I failed 🤣 but given that I really like this idea I'll invest some time during the weekend :D 

btw. would be great if you could share the example module already that excludes... maybe `ps_imageslider` or something visible, that would help QA

**Comment by kpodemski** [comment]
btw, @matthieu-rolland - @Codencode has already written a proposed solution based on my other suggestion:
https://github.com/PrestaShop/PrestaShop/discussions/37060#discussioncomment-10841735

I haven't tested it and I haven't gotten back to him with a review (because I see room for improvement), but this could give you an idea how to do that w/o messing with the front container

**Comment by matthieu-rolland** [comment]
I have issues in integration and unit tests, clearly related to my PR, I'm on it...

**Comment by matthieu-rolland** [comment]
> btw, @matthieu-rolland - @Codencode has already written a proposed solution based on my other suggestion: [#37060 (comment)](https://github.com/PrestaShop/PrestaShop/discussions/37060#discussioncomment-10841735)
> 
> I haven't tested it and I haven't gotten back to him with a review (because I see room for improvement), but this could give you an idea how to do that w/o messing with the front container

ah thanks, interesting 👍   fyi I'll be off for two weeks starting this evening, I'll see @jolelievre 's feedback too (I discussed this with him too)

File: src/Adapter/ContainerFinder.php
Comment (matthieu-rolland) [inline src/Adapter/ContainerFinder.php]: Actually, the $container variable is always set, it never fails...

And I think it could be the reason for the broken unit and integration test, seems like at some point we're fetching services on the wrong container ?

**Comment by Codencode** [comment]
> ah thanks, interesting 👍 fyi I'll be off for two weeks starting this evening, I'll see @jolelievre 's feedback too (I discussed this with him too)

Hello @matthieu-rolland,
I also created a small module to test the functionality. The module blocks the displayFooter hook for the ps_linklist module and the displayFooterBefore hook for the ps_emailsubscription module.

You can always find the code here:
https://github.com/PrestaShop/PrestaShop/discussions/37060#discussioncomment-10841735

**Comment by kpodemski** [comment]
@matthieu-rolland I managed to fix the CI. It's a bit dirty because we get the container explicitly, but the code you modified should already get the front container when needed. It could be because we first instantiated it using the static method. I'm not sure. Anyway, CI is green, I'm attaching example module, I tested it and it works as expected.

[kpfiltermodules.zip](https://github.com/user-attachments/files/17350039/kpfiltermodules.zip)

**Comment by matthieu-rolland** [comment]
> @matthieu-rolland I managed to fix the CI. It's a bit dirty because we get the container explicitly, but the code you modified should already get the front container when needed. It could be because we first instantiated it using the static method. I'm not sure. Anyway, CI is green, I'm attaching example module, I tested it and it works as expected.
> 
> [kpfiltermodules.zip](https://github.com/user-attachments/files/17350039/kpfiltermodules.zip)
> 
> 
> 

You rock !
 hopefully it can be merged before I come back (this PR really needs @jolelievre 's approval before merging though, I made the container available following his recommendation and I want to be sure he is ok with this !)

Thanks! See you in two weeks

**Review by tleon — DISMISSED** [review]

File: config/services/front/services_prod.yml
Comment (jolelievre) [inline config/services/front/services_prod.yml]: You're not supposed to need this, the container builder (even in front) already loads dynamically all the config files in `modules/*/config/front/services.yml` Seems like you are introducing a new file `from_services.yml` that is not necessary and could lead to confusion because it means two different files can be used

Comment (jolelievre) [reply config/services/front/services_prod.yml]: Besides if you implement it this way you will only allows filtering hooks in front, so only in legacy but it won't work for back office which load different config files `modues/*/config/admin/services.yml`

So to make this work in both environment you should define you filtering services in `modules/*/config/services.yml` which is the common service file for all environements

**Review by jolelievre — CHANGES_REQUESTED** [review]

File: config/services/front/services_prod.yml
Comment (matthieu-rolland) [reply config/services/front/services_prod.yml]: done 👍 
Indeed you're right

File: classes/Hook.php
Comment (jolelievre) [inline classes/Hook.php]: Hmm, it's surprising that you only fetch the container for front 🤔 The fetched container should adapt on the environment you are currently in, in BO the admin container (Symfony one) in legacy environment (like the FO) the legacy one based on `Adapter\ContainerBuilder`

This code is executed in all environments, and it's at the core of the Hook execution which is used pretty much everywhere, so with the current code it you are forcing the creation and build of the Front container even when we are in the BO which is not needed and could create some unexpected side effects

Comment (jolelievre) [reply classes/Hook.php]: You should use `SymfonyContainer::getInstance` first, and when it returns null it means you are in an FO environment so you could use the legacy COntainerBuilder as a fallback in this case

It also means something very important, the filter service you introduced along with the interfaces implemented in the modules must be present in ALL containers (both legacy and symfony) meaning you shouldn't define them in the front services only but in the common ones https://github.com/PrestaShop/PrestaShop/blob/e8c31d70a6d52ed235a60caf673c98a5db08a029/config/services/common.yml

Finally, you only declared your service in the `prod` which means your feature won't work when the shop is in dev or test mode

Fro what I remember your filter feature should be needed everywhere not only in the FO environment and only in prod, so you need all these adjustments to make your implementation more generic

File: src/Core/Hook/HookModuleFilter.php
Comment (jolelievre) [inline src/Core/Hook/HookModuleFilter.php:41]: I keep struggling on this class name, the service is not a filter itself It's the service responsible for filtering so it should be something closer to `HookModuleFilterer` but I believe it's not really a word 😅

Comment (jolelievre) [reply src/Core/Hook/HookModuleFilter.php:41]: What is the purpose of this implementation? Remove some hook, update some hook, add some hook?

**Review by jolelievre — CHANGES_REQUESTED** [review]

File: src/Core/Hook/HookModuleFilter.php
Comment (matthieu-rolland) [reply src/Core/Hook/HookModuleFilter.php:41]: discussed on slack

File: classes/Hook.php
Comment (matthieu-rolland) [reply classes/Hook.php]: I dealt with those feedbacks 👍  (except about the `prod` config, the dev config actually extends the prod one...)

File: src/Core/Hook/HookModuleFilter.php
Comment (jolelievre) [inline src/Core/Hook/HookModuleFilter.php:41]: The naming of the class is not ideal, so you should at least add a description about the purpose of this service what it does, an example of a use case and maybe explain a little that it is based on tags so modules can add their own implementation

This is especially useful because there is no use case in the core itself, the actual implementation will be in external modules so a future developer has no way to understand why this is needed

File: src/Core/Hook/HookModuleFilterInterface.php
Comment (jolelievre) [inline src/Core/Hook/HookModuleFilterInterface.php:33]: Same suggestion this deserves some explanation

**Review by jolelievre — COMMENTED** [review]
Thanks @matthieu-rolland 

Seems good to me, except it deserves some additional comment to understand the purpose of this feature, and the tests are red but I don't know if it's related to your modif

File: config/services/common.yml
Comment (jolelievre) [inline config/services/common.yml:31]:
```suggestion
```
Since the service is already defined in `../../src/PrestaShopBundle/Resources/config/services/core/common.yml` which is loaded in this file you don't need to duplicate the definition here

Comment (matthieu-rolland) [reply config/services/common.yml:31]: it doesn't work anymore in FO if I remove this service definition... the service `PrestaShop\PrestaShop\Core\Hook\HookModuleFilter` is unfound

**Review by jolelievre — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]
Accepted but please either provide a test module or update mine.

**Comment by kpodemski** [comment]
ping @matthieu-rolland ?

**Review by boherm — APPROVED** [review]

**Comment by matthieu-rolland** [comment]
test module added in description 👍

**Comment by paulnoelcholot** [comment]
Hello @matthieu-rolland,

I tested your PR and it's good for me! :tada: 

Thanks!

Auto test : https://github.com/paulnoelcholot/testing_pr/actions/runs/12374859029

image

**Review by paulnoelcholot — APPROVED** [review]

**Comment by matthieu-rolland** [comment]
Thank you !

**Comment by Codencode** [comment]
@matthieu-rolland, @kpodemski,
I've been reviewing this PR and noticed an issue that I’d like to bring to your attention.

I’ve noticed that this change causes a significant slowdown in the back office.
From what I observed, the issue lies in the method `HookCore::getHookModuleFilter()`, where the system attempts to retrieve the service `PrestaShop\PrestaShop\Core\Hook\HookModuleFilter` using the following instruction:
```php
try {
    $hookModuleFilter = $serviceContainer->get(HookModuleFilter::class);
} catch (Exception $e) {
    return null;
}
```
However, in the back office, the system fails to retrieve it and throws the following exception (handled via try/catch, so it’s not visible):


```php
throw new ServiceNotFoundException($id, null, null, [], sprintf('The "%s" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.', $id));
```
This exception is found in the method`Symfony\Component\DependencyInjection\Container::make()`

In the front office, the issue does not occur because the service is retrieved correctly.
The slowdown is not negligible since the exception is thrown approximately 1,000 times per request, leading to the slowdown.

The call stack is as follows:

- PrestaShop\PrestaShop\Adapter\Hook\HookDispatcher::__construct()
- Symfony\Component\EventDispatcher\EventDispatcher::addSubscriber()
- PrestaShop\PrestaShop\Adapter\LegacyHookSubscriber::getSubscribedEvents()
- HookCore::getHookModuleExecList() https://github.com/PrestaShop/PrestaShop/blob/2a02d2fa63c1576de909a696ce60d0dc996fc829/classes/Hook.php#L866-L868
- HookCore::getHookModuleFilter() https://github.com/PrestaShop/PrestaShop/blob/2a02d2fa63c1576de909a696ce60d0dc996fc829/classes/Hook.php#L1307C1-L1311C10
- Symfony\Component\DependencyInjection\Container::get()
- Symfony\Component\DependencyInjection\Container::make(): _the exception is thrown here_ https://github.com/symfony/dependency-injection/blob/7a379d8871f6a36f01559c14e11141cc02eb8dc8/Container.php#L248C1-L250C14

I hope I’ve explained this clearly enough.
I would like some feedback to understand if this issue is occurring only for me or if there’s something that needs to be fixed.

Thank you!

**Comment by Codencode** [comment]
From what I understand, the issue is that the system cannot find the PrestaShop\PrestaShop\Core\Hook\HookModuleFilter service, so I tried setting public: true in the service definition like this:

```yml

  PrestaShop\PrestaShop\Core\Hook\HookModuleFilter:
    public: true
    arguments:
      - !tagged core.hook_module_exec_filter
 ```
in this file:

https://github.com/PrestaShop/PrestaShop/blob/2a02d2fa63c1576de909a696ce60d0dc996fc829/src/PrestaShopBundle/Resources/config/services/core/common.yml#L15-L17

By doing so, the exception is no longer thrown, and the slowdown is drastically reduced.

I’m not sure if this is the solution to the problem, but I wanted to share this test to help identify the potential issue.

**Comment by matthieu-rolland** [comment]
Thank you @Codencode for your feedback !
I'll find some time and check this next week👍

**Comment by matthieu-rolland** [comment]
@Codencode here's the fix PR: https://github.com/PrestaShop/PrestaShop/pull/37912

thank you very much for this nice catch !

**Comment by Codencode** [comment]
> [matthieu-rolland](/matthieu-rolland)

Great @matthieu-rolland, thanks to you!

**Pull request merged**
